node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -125,12 +126,15 @@ retrievalWorkerImpl SendActions {..} =
                 let firstHeader = headers ^. _Wrapped . _neLast
                 in handleCHsValid enqueueMsg endedRecoveryVar nodeId
                                   firstHeader (headerHash header)
-        void $ enqueueMsg (MsgRequestBlockHeaders (Just (S.singleton nodeId))) $ \_ _ -> pure $ Conversation $ \conv ->
+        convs <- enqueueMsg (MsgRequestBlockHeaders (Just (S.singleton nodeId))) $ \_ _ -> pure $ Conversation $ \conv ->
             requestHeaders cont mgh nodeId conv
                 `finally` void (tryPutMVar endedRecoveryVar False)
-        -- Block until the conversation has ended.
-        whenM (readMVar endedRecoveryVar) $
```

Comment:
Without this patch, a blocked indefinitely on MVar exception would be raised here in case no peer is reached (the conversation body never runs).